storage/innobase/buf/buf0buf.cc

```diff
@@ -838,10 +839,11 @@ class mem_pressure
   void trigger_collection()
   {
     uint64_t u= 1;
-    if (m_event_fd >=0 && write(m_event_fd, &u, sizeof(uint64_t)) != sizeof(uint64_t))
+    if (m_event_fd < 0 || write(m_event_fd, &u, sizeof(uint64_t)) != sizeof(uint64_t))
+    {
       sql_print_information("InnoDB: (Debug) Failed to trigger memory pressure");
-    else /* assumed failed to meet intialization criteria, so trigger directy */
       buf_pool.garbage_collect();
+    }
   }
```

Comment:
Aha. So this is the reason that the existing test case did not fail from the bug. There were actually two bugs, the inverted m_abort test, and this incorrect test which (IIUC) ran the garbage_collect() when it _succeeded_ in triggering the memory pressure thread, rather than when it failed to trigger (as I assume was intended?).

So the two bugs kind of cancelled each other, and the garbage_collect() was triggered anyway.

But why have this condition at all? What is the purpose of triggering buf_pool.garbage_collect() directly here if you cannot signal the thread? That seems to only hide bugs (as indeed it did in this case).

So better just remove this call of garbage_collect()?

Reply:
So now with this change, you have a race condition in the test case. Because the test case checks for message from the garbage_collect() in the error log, but this message only comes asynchroneously, and thread scheduling can delay the mem_pressure thread and make the test not see the message and fail.

You can easily see this by putting a my_sleep(1000000) in the thread routine just before running the garbage_collect(), then the test fails.

This is going to trigger random failures in buildbot, so needs to be fixed.

Reply:
yep - `garbage_collect` removed and replaced with a mtr test case for cgroupsv2 only.

Race condition fixed.